virttest/graphical_console.py

```diff
@@ -31,22 +34,76 @@ def get_keymap(keymap_file):
         return keymap_dict
 
 
+def uniform_linear(duration, rate):
+    """
+    Uniform linear trace for mouse move.
+
+    :param duration: the move duration time.
+    :param rate: frequency of sending move event.
+    :param src: start position.
+    :param dst: destination position.
+
+    :return: the object with a position in this line.
+    :rtype: function for motion.
+    """
+
+    def motion(src, dst):
+        x1, y1 = src
+        x2, y2 = dst
+        x, y = x1, y1
+        num = duration * rate
+        dx_int, dx_float = str((x2 - x1) / num).split('.')
+        dy_int, dy_float = str((y2 - y1) / num).split('.')
+        decimal_x = 0
+        decimal_y = 0
+        for i in range(num):
+            decimal_x += float("0." + dx_float)
+            decimal_y += float("0." + dy_float)
+            if decimal_x >= 1:
+                x += (-1 if "-" in str(dx_int) else 1)
+                decimal_x = float("0." + (str(decimal_x).split(".")[1]))
+            if decimal_y >= 1:
+                y += (-1 if "-" in str(dy_int) else 1)
+                decimal_y = float("0." + (str(decimal_y).split(".")[1]))
+            x, y = x + int(dx_int), y + int(dy_int)
+            if i == num - 1:
+                x, y = x2, y2
+            yield (x, y)
+            time.sleep(duration / rate)
+    return motion
+
+
 class UnsupportedKeyError(Exception):
     """
     Unsupported Key Error for Console
     """
     pass
 
 
+class ConsoleNotSupportedError(Exception):
+    """
+    Console not supported error.
+    """
+    pass
+
+
 class BaseConsole(object):
     """
     Base Console
     """
     KEY_DOWN = 1
     KEY_UP = 0
 
+    BTN_DOWN = 1
+    BTN_UP = 0
+
+    SCROLL_FORWARD = 1
+    SCROLL_BACKWARD = -1
+
     def __init__(self, vm, logfile=None):
         self._vm = vm
+        self.width, self.height = self.screen_size()
+        self.pointer_pos = (int(self.width/2), int(self.height/2))
```

Comment:
Can't assume the pointer always be at the centre of the screen, for example, when you reattach the console with a newly created object, the pointer may be away of the centre.

Reply:
I do a pointer position initialization here, if users create a new graphical console object, the default pointer position will set to this console screen central, it will be changed along with pointer move, users will always get the latest pointer position in the same console, but if they create a new object again, the pointer position will also be re-initiated in this new console.

Reply:
Let me explain about that, supposes that you have created an console object and moved the pointer to somewhere, for example the top-right, the screen would look like the following (`^` is presented as the pointer):
```
+-------+
|     ^ |
|       |
|       |
+-------+
```
then if you close the console connection and recreate a new one, by the current approach it would look like:
```
  +-------+
+ | - - + |
  |   ^   |
| |     | |
  +-------+
+ - - - +
```
the box drawn by solid line is the screen seen by the current console, and the one drawn by dash line is the screen seen by the previous, they do not coincide, and that means you have a new coordinate system now, then the question is which one should you choose to get the position of the pointer?

Reply:
as we discussed, will use a class attribute to save pointer position, will get pointer position dynamically through this class attribute. Thanks a lot~

Reply:
Updated the codes, new added a class method _set_pointer_pos() to change the class attribute POINTER_POS, because instance method cannot change it.
Thanks~